### ISSUE TITLE
[heartbeat] A heart beat p2p message to broadcast the node info to the network

### ISSUE THREAD

**Issue by LeoHChen** [harmony-one/harmony#3370]
# Problem/limitation At Hand

We don't have a way to collect the external nodes' data at the network scale. Even though we have an RPC interface, the RPC interface is only useable from the localhost by default, which is the right thing to do for security reasons.

We need to collect network-level stats on the node's info for better monitoring.

**Which component?**

- [x] p2p

# Proposed Solution

This feature request is to add a new p2p message to the node, to expose some node data to the network. All the data we collect should be non-sensitive data for security reasons. 
The heartbeat message can be broadcast to the network per minute to avoid the impact of the consensus.

The essential data we would like to collect includes, but not limited to.

- validator node info
  - node version
  - number of blskeys
  - list of blskeys
  - validator account
  - node type: validator/explorer

- host info
  - OS version
  - size of RAM
  - number of CPU, and average/peak load
  - cloud provider

- performance metrics
  - number of consensus
  - average finality time

# Collection
We can use our snapshot or internal explorer nodes to collect the heartbeat data.  The data can be exposed to Prometheus and present using Grafana.

**Comment by JackyWYX** [comment]
Are we adding a new topic?

**Comment by LeoHChen** [comment]
yes, this can be a new topic.

**Comment by JackyWYX** [comment]
My design suggestion are as follows:

### 1. Encapsulation

Encapsulate all heart beat code to one or two object. So that instead of adding more code to "god class" node, we can simply add one object to it.
The object would be better to have a well defined interface like

```
// go-ethereum/node/lifecycle.go:23
type Lifecycle interface {
    // Start is called after all services have been constructed and the networking
    // layer was also initialized to spawn any goroutines required by the service.
    Start() error

    // Stop terminates all goroutines belonging to the service, blocking until they
    // are all terminated.
    Stop() error
}
```

so that higher-level class node does not need to know the implementation of the heartbeat client (host) service. You can take a look how Ethereum use the interface to simplify the code logic (node/node.go:163).

### 2. Components providing heartbeat data

I think define a dataProvider interface in heartbeatClient for components would be a good practice:

```
// network/types.go
type DataPack interface {
    Encode() []byte
    Decode([]byte) 
}

// heartbeatClient.go
type heartbeatDataProvider interface {
    getHeartbeatData() (nettypes.DataPack, error)
}

type heartbeatClient struct {
    providers map[string]heartbeatDataProvider
}

func (client *heartbeatClient) getHeartbeatData() map[string][]byte {
    ...
}
```

The `heartbeatDataProvider` interface can be implemented by all major components: p2p, consensus (validator + leader), sync, rpc (or rosetta), blockchain (aka core), misc (version, e.t.c). Also these metrics can be also used for prometheus monitoring.

Of course, if we don't want to add that much data to heart beat message, it's totally fine. But I believe this design will give more extensibility to the system we are building :)

**Comment by yelllowsin** [comment]
Thanks guys for going forward on implementing this, super requested/important feature to track the nodes version on the network. Some comments:

1) Can we have the "IP Address" as in the **host info** section? So we can track the geo location of nodes around the globe and draw a map on our dashboards. I think this is more useful than the Cloud Provider field, which i am not sure is that trustworthy/accurate? I don't see this being used on other blockchains, usually they rely on the IP address to track the geo location.

2) About exposing data to Prometheus and presenting using Grafana, we have a bounty open with @Chainode where he proposed to created this on the `harmony.conf` file. He would then created the Grafana dashboard template and the Prometheus + Grafana setup for easy node monitoring.

```
[Prometheus]
Prometheus = true
Prometheus_listen_addr=":Port"
```

Parameters to be monitored via Prometheus enabled flag:

- Signing on each node;
- Shard for each node;
- Harmony SW version on each node;
- Block height on each node;
- Election state;
- Delegation amount;
- Nr. of connected peers;
- Block time on each shard;
- .....(to be expanded)

**Comment by pops-one** [comment]
@yelllowsin no issue with everything you said except on the IP. I believe it will be up to the validator willingness to disclose his IP or not. IP can be scanned/attacked (if any vulnerabilities, or brute forced) so this is a double edge sword. 
If we wanna push for the IP, maybe go and mask the ip last octet ie (35.12.98.xxx) at least there are 255 other IPs to be brute forced/scanned.  

My take here is to not share at all the IP.

**Comment by LeoHChen** [comment]
I've decided to use prometheus framework to collect the metrics.
https://github.com/harmony-one/harmony/pull/3454

**Comment by LeoHChen** [comment]
prometheus metrics is supported and released in v3.0.0.
